keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +336,169 @@ or later. The dual-stack feature requires no changes to this API.
+
+1. The maximum allowed number of cidrs in this flag is two. Attempting to use more than two cidrs will generate failures during startup. 
+2. When configured with two cidrs then the values are expected to have different ip families. e.g `--services-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>` or `--services-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`. Attempting to configure a cluster with two similar cidrs will generate failures during startup.
+3. A dualstack cluster can have a single cidr in this flag. That means the cluster will be to egress traffic on both ip families, but will *not* be ingress traffic on both ip families.
+4. When upgrading a cluster to dualstack, the first element of the `--services-cluster-ip-range` flag MUST NOT change IP families.  As an example, an IPv4 cluster can safely be upgraded to use `--services-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>`, but can not safely be changed to `--services-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`.
+5. The first cidr entry of `--services-cluster-ip-range=<first cidr entry>, <secondary cidr entry>` is considered cluster default address family.
+6. Attempting to allocate `ClusterIP` from an ip address family that is not configured on the cluster will result in an error.
+
+> The below discussion assumes that `Endpoint Slice and Endpoint Slice Controller`  will allow dual stack endpoints, the existing endpoint controller will remain single stack. That means enabling dual stack services by having more than one entry in `--services-cluster-ip-range` without enabling `Endpoint Slice` feature gate will result in validation failures during startup.
 
 #### Type NodePort, LoadBalancer, ClusterIP
-Services will only be assigned an address from a single address family (either IPv4 or IPv6). The address family for the Service’s cluster IP is determined by setting the field, `.spec.ipFamily`, on that Service. You can only set this field when creating a new Service. Setting the `.spec.ipFamily` field is optional and should only be used if you plan to enable IPv4 and IPv6 Services and Ingresses on your cluster. The configuration of this field not a requirement for egress traffic. The default address family for your cluster is the address family of the first service cluster IP range configured via the `--service-cluster-ip-range` flag to the kube-controller-manager. The `.spec.ipFamily` field can be set to either IPv4 or IPv6.
+In a dualstack cluster Services can be:
+1. Singlestack: assigned from the first or the second cidr entries as set using `--services-cluster-ip-range` flag, in this scenario Services will have a single allocated `ClusterIP`
+2. dualstack: assigned from both the first and the second cidr entries, in this scenario Services will have a dual assigned `ClusterIPs`
+
+The above is achieved using the following changes to Service api.
+1. Add `spec.ipFamilyPolicy` field with possible values of  `SingleStack`, `DualStack`, or `nil`.
```

Comment:
I saw the earlier comment. I want to have this field as a string because it allow us to do creative things in the future when services have >2 IPs. i.e. *all available IPv4 on pods* or something like that, no concrete use case in mind yet. Also `*bool` works with golang, I am not sure it works with other languages where pointers are not a thing.

Reply:
We have lots of `*bool` already

    danw@p50:api (master)> grep -c '\*bool' core/v1/types.go 
    22

Though for most of those it's not clear that unset and explicitly-`false` are any different...

Reply:
exactly my concern.  I am ok with this as an enum, let's just think about whether "policy" is the best name, then

Reply:
Fwiw, we have examples of `*bool` ternaries where "unset" means something different to true or false.  Eg: `automountServiceAccountToken` - unset means cascade up Pod->ServiceAccount->true.  All other languages have had to come up with some way to represent this too (eg: `Option<bool>` in Rust or `True`/`False`/`None` in python)